docs/submitting-applications.md

```diff
@@ -48,13 +48,22 @@ Some of the commonly used options are:
 * `application-jar`: Path to a bundled jar including your application and all dependencies. The URL must be globally visible inside of your cluster, for instance, an `hdfs://` path or a `file://` path that is present on all nodes.
 * `application-arguments`: Arguments passed to the main method of your main class, if any
 
-<b>&#8224;</b> A common deployment strategy is to submit your application from a gateway machine
-that is
-physically co-located with your worker machines (e.g. Master node in a standalone EC2 cluster).
-In this setup, `client` mode is appropriate. In `client` mode, the driver is launched directly
-within the `spark-submit` process which acts as a *client* to the cluster. The input and
-output of the application is attached to the console. Thus, this mode is especially suitable
-for applications that involve the REPL (e.g. Spark shell).
+<b>&#8224;</b>In `client` mode, the driver is launched directly within the `spark-submit` process on
+the machine which was used to submit the Spark job, and it will act as a *client* to the cluster.
```

Comment:
eg. in client mode for k8s, driver might not be in the spark-submit process but instead in a pod in the cluster
https://spark.apache.org/docs/latest/running-on-kubernetes.html#client-mode

Reply:
OK, maybe best to punt on this.

Reply:
I have added some further description, now Kubernetes docs is also linked.

Reply:
I actually think that distinction makes things less clear.

client mode = driver runs in same machine as spark-submit
cluster mode = driver runs in some cluster machine

It doesn't matter what machine it is that is running spark-submit; it may be a cluster machine, a k8s pod, whatever. The distinction is whether the driver runs in the same machine as spark-submit or not.

Reply:
well, that's terminology perhaps we need to change for k8s.

Reply:
Why so? What I wrote applies to k8s too.

In your original comment above, you mention that the driver might be in a pod in the cluster. But that pod it executing spark-submit. So it's exactly the same situation.

Reply:
Guys, could you please come up with an agreement on the wording? I am happy to update the PR if you put your preferred version to here as comment. 
The current description is not detailed enough and apparently is not applicable to some of the cases: that should be improved...

Reply:
I already explained above the different between client and cluster mode. I haven't seen anything that tells me that what I wrote is not correct in all cases.

If you run spark-submit in client mode, the driver is run by spark-submit. No cluster resources are requested for the driver.

If you run spark-submit in cluster mode, spark-submit will ask the cluster manager to run the driver. Resources will be allocated from the cluster manager to run the driver, which will then run in a different place from spark-submit itself.

There are no differences among the cluster managers here. The fact that you can run spark-submit in client mode inside a cluster node is irrelevant. From the standpoint of spark-submit, the above still applies.

Reply:
I think I agree with @vanzin here. The comment above is accurate, and probably as far as we need to distinguish it for purposes of what this comment is trying to explain. I'd propose removing the special comment about K8S.

I might phrase the distinction as:

- In `client` mode, the `spark-submit` process runs the driver directly, from the machine running `spark-submit`. That process is the direct 'client' to the cluster workers running the executors, and runs as long as the job does. If this `spark-submit` process is terminated, the driver terminates and thus the job does too.
- In `cluster` mode, the `spark-submit` process runs the driver on the cluster. That process and machine is not involved in running the job after submission. The job continues to run even if `spark-submit` terminates, as the driver 'client' is on the cluster. The job runs as long as the driver continues to execute on the cluster.

That kind of thing.